### PR TITLE
StringLowering: Escape the JSON in the custom section

### PR DESCRIPTION
**Pull request by kripken**
Also add an end-to-end test using node to verify we can parse the escaped
content properly using TextDecoder+JSON.parse.

### CODE REVIEW

File: src/support/string.h [hunk 1]
```diff
@@ -75,7 +75,14 @@ inline bool isNumber(const std::string& str) {
   return !str.empty() && std::all_of(str.begin(), str.end(), ::isdigit);
 }
 
-std::ostream& printEscaped(std::ostream& os, std::string_view str);
+enum EscapeMode {
+  Normal,
+  JSON
+};
+
+std::ostream& printEscaped(std::ostream& os,
```
Comment (tlively) [inline src/support/string.h:83]: I think having a separate `printJSONEscaped` function would make more sense for this interface, even if the implementation still uses a flag.

File: src/support/string.cpp [hunk 2]
```diff
@@ -132,7 +132,13 @@ std::ostream& printEscaped(std::ostream& os, std::string_view str) {
         if (c >= 32 && c < 127) {
           os << c;
         } else {
-          os << std::hex << '\\' << (c / 16) << (c % 16) << std::dec;
+          if (mode == EscapeMode::Normal) {
+            os << std::hex << '\\' << (c / 16) << (c % 16) << std::dec;
+          } else if (mode == EscapeMode::JSON) {
+            os << std::hex << "\\u00" << (c / 16) << (c % 16) << std::dec;
```
Comment (tlively) [inline src/support/string.cpp:138]: JSON requires exactly 4 hex digits representing a UTF-16 code unit (i.e. 2 bytes at once), so it's going to be slightly more complicated than this, unfortunately.

**Comment by kripken** [comment]
PR now contains code point logic from Emscripten, that is hopefully correct... on a large real-world testcase it seems to parse everything properly at least.

**Review by tlively — COMMENTED** [review]
The encoding looks right now, but only as long as the string is valid WTF-8. At no point do we validate or otherwise require that strings are valid WTF-8 (although perhaps we should be validating this?), so I think it would make sense to code more defensively here.

We could also make this code much more readable :)

It would be good to get more test coverage as well.

**Comment by kripken** [comment]
I ported the warning from the JS code to be more defensive here, and I added more comments in general. Did you have anything else in mind for defense/readability?

I added test coverage for all escaped characters.

**Comment by kripken** [comment]
Last commit adds a test for a weird utf8 char, and fixes our handling of the escape codes.

File: src/support/string.cpp [hunk 3]
```diff
@@ -164,44 +164,58 @@ std::ostream& printEscapedJSON(std::ostream& os, const std::string_view str) {
...
+        // This uses 3 bytes.
         i++;
         int u2 = str[i] & 63;
         if ((u0 & 0xF0) == 0xE0) {
           u0 = ((u0 & 15) << 12) | (u1 << 6) | u2;
         } else {
+          // This uses 4 bytes.
+          if ((u0 & 0xF8) != 0xF0) {
+            std::cerr << "warning: Bad UTF-8 leading byte " << int(u0) << '\n';
+          }
```
Comment (tlively) [inline src/support/string.cpp:210]: It would be good to emit similar warnings in the 1-, 2-, and 3-byte cases as well

**Comment by tlively** [comment]
I plan to try to rewrite this to be safer and easier to understand, so LGTM as-is for the short term.

**Pull request merged**